block-producer/src/batch_builder/batch.rs

```diff
@@ -64,14 +66,38 @@ impl TransactionBatch {
         let produced_nullifiers =
             txs.iter().flat_map(|tx| tx.input_notes().iter()).cloned().collect();
 
-        let (created_notes, created_notes_smt) = {
-            let created_notes: Vec<NoteEnvelope> = txs
+        let (created_notes, created_notes_smt, created_note_envelopes) = {
```

Comment:
I think we shouldn't produce `NoteCreated` here, since we don't know all needed fields at the moment (we don't know `batch_index`). We can provide note envelopes here and serialized details and construct `NoteCreated` in the block builder.

So I would rewrite this, something like:
```rust
        let (created_notes_smt, created_note_envelopes) = {
            let created_note_envelopes: Vec<(NoteEnvelope, Option<Vec<u8>>)> = txs
                .iter()
                .flat_map(|tx| tx.output_notes().iter())
                .map(|note| {
                    match note {
                        OutputNote::Public(note) => (note.into(), Some(note.to_bytes())),
                        OutputNote::Private(envelope) => (*envelope, None),
                    }
                })
                .collect();

            if created_note_envelopes.len() > MAX_NOTES_PER_BATCH {
                return Err(BuildBatchError::TooManyNotesCreated(created_note_envelopes.len(), txs));
            }

            // TODO: document under what circumstances SMT creating can fail
            (
                BatchNoteTree::with_contiguous_leaves(
                    created_note_envelopes
                        .iter()
                        .map(|(note_envelope, _)| (note_envelope.id(), note_envelope.metadata())),
                )
                .map_err(|e| BuildBatchError::NotesSmtError(e, txs))?,
                created_note_envelopes,
            )
        };

        Ok(Self {
            id,
            updated_accounts,
            produced_nullifiers,
            created_notes_smt,
            created_notes,
        })
```